build_tools/bazel_to_cmake/bazel_to_cmake.py

```diff
@@ -328,8 +329,9 @@ def convert_directory(directory_path, write_files, allow_partial_conversion, ver
         preserved_header + header + converted_build_file + preserved_footer
     )
     if write_files:
-        with open(cmakelists_file_path, "wt") as cmakelists_file:
-            cmakelists_file.write(converted_content)
+        with tempfile.NamedTemporaryFile("wt", delete=False) as tmp:
+            tmp.write(converted_content)
+        os.replace(tmp.name, cmakelists_file_path)
```

Comment:
This fails on my Windows machine since I have my repository on a different disk drive than my temp location:

```
Using repo root D:\dev\projects\iree
Traceback (most recent call last):
  File "D:\dev\projects\iree\build_tools\bazel_to_cmake\bazel_to_cmake.py", line 395, in <module>
    main(parse_arguments())
  File "D:\dev\projects\iree\build_tools\bazel_to_cmake\bazel_to_cmake.py", line 361, in main
    convert_directories(
  File "D:\dev\projects\iree\build_tools\bazel_to_cmake\bazel_to_cmake.py", line 196, in convert_directories
    status = convert_directory(
             ^^^^^^^^^^^^^^^^^^
  File "D:\dev\projects\iree\build_tools\bazel_to_cmake\bazel_to_cmake.py", line 334, in convert_directory
    os.replace(tmp.name, cmakelists_file_path)
OSError: [WinError 17] The system cannot move the file to a different disk drive: 'C:\\Users\\Scott\\AppData\\Local\\Temp\\tmpwbz6zgps' -> 'compiler\\src\\iree\\compiler\\Dialect\\Stream\\Conversion\\FlowToStream\\CMakeLists.txt'
```

`shutil.move` recommended on https://stackoverflow.com/a/21116654 works around that. Does that still have the atomic properties that will hopefully fix the issue?

Reply:
I don't see any atomicity guarantee at https://docs.python.org/3/library/shutil.html#shutil.move. So probably no.

Your case is probably common, though; I wonder what's the normal way to handle that. Maybe @benvanik knows?

Reply:
Maybe we should create the temp file in the same directory as the destination?
https://stackoverflow.com/questions/8745387/in-python-make-a-tempfile-in-the-same-directory-as-another-file

Reply:
Pushed a commit doing that as in the SO thread.

Reply:
Oof. That looks risky, and testing it confirms:

```
Run bazel_to_cmake.py....................................................Failed
- hook id: bazel_to_cmake
- exit code: 1

Using repo root D:\dev\projects\iree
0 CMakeLists.txt files were updated, 1 were skipped, and 10 required no change.
Using repo root D:\dev\projects\iree
0 CMakeLists.txt files were updated, 1 were skipped, and 11 required no change.
Using repo root D:\dev\projects\iree
0 CMakeLists.txt files were updated, 0 were skipped, and 10 required no change.
Using repo root D:\dev\projects\iree
Traceback (most recent call last):
  File "D:\dev\projects\iree\build_tools\bazel_to_cmake\bazel_to_cmake.py", line 404, in <module>
    main(parse_arguments())
  File "D:\dev\projects\iree\build_tools\bazel_to_cmake\bazel_to_cmake.py", line 370, in main
    convert_directories(
  File "D:\dev\projects\iree\build_tools\bazel_to_cmake\bazel_to_cmake.py", line 197, in convert_directories
    status = convert_directory(
             ^^^^^^^^^^^^^^^^^^
  File "D:\dev\projects\iree\build_tools\bazel_to_cmake\bazel_to_cmake.py", line 345, in convert_directory
    os.replace(tmp.name, cmakelists_file_path)
PermissionError: [WinError 5] Access is denied: 'D:\\dev\\projects\\iree\\compiler\\src\\iree\\compiler\\Dialect\\Stream\\IR\\test\\CMakeLists.txt1oau1aa8' -> 'compiler\\src\\iree\\compiler\\Dialect\\Stream\\IR\\test\\CMakeLists.txt'
```

```
λ git status
On branch local-address-fix
You are currently bisecting, started from branch 'bisect-releases'.
  (use "git bisect reset" to get back to the original branch)

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   build_tools/bazel_to_cmake/bazel_to_cmake.py

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        compiler/src/iree/compiler/Bindings/Native/CMakeLists.txt7fd6d0_v
        compiler/src/iree/compiler/Dialect/Stream/IR/test/CMakeLists.txt1oau1aa8

no changes added to commit (use "git add" and/or "git commit -a")
```

Reply:
I don't have a mental model that helps me understand why you knew it was risky, and why it actually resulted in access denied. Bowing out!

Reply:
Risky in that it writes into the source tree, and if there are failures or the process is interrupted then there will be orphaned temp files. We could maybe write into a temp folder on the source _drive_, if those APIs support that.

Thanks for the change proposals! I can send a PR that splits the pre-commit hook into two and we can see if that helps.

Reply:
I was kind of guessing that part of the risk, but what's mysterious to me is why did you get access denied?

Reply:
No idea 😅. I would have expected that to work. Ran the command between meetings to see but didn't dig any deeper.